crates/ingestion-client/src/session.rs

```diff
@@ -0,0 +1,507 @@
+                            // speed up over a period of time.
+
+                            debug!("Ingestion error from {}: {}", connection.peer(),  err);
+                            break SessionState::Disconnected;
+                        }
+                    }
+                }
+            }
+        };
+
+        // state == Disconnected
+        assert!(matches!(state, SessionState::Disconnected));
+
+        // don't lose the buffered batch
+        let remainder = chunked.into_remainder();
+        if !remainder.is_empty() {
+            self.inflight.push_back(IngressBatch::new(remainder));
+        }
+
+        state
+    }
+}
+
+struct SessionManagerInner<T> {
+    networking: Networking<T>,
+    partition_routing: PartitionRouting,
+    opts: SessionOptions,
+    cancellation: CancellationToken,
+    handles: DashMap<PartitionId, SessionHandle>,
+}
+
+impl<T> SessionManagerInner<T>
+where
+    T: TransportConnect,
+{
+    /// Gets or start a new session to partition with given partition id.
+    /// It guarantees that only one session is started per partition id.
+    pub fn get(&self, id: PartitionId) -> SessionHandle {
+        self.handles
+            .entry(id)
+            .or_insert_with(|| {
+                let session = PartitionSession::new(
+                    self.networking.clone(),
+                    self.partition_routing.clone(),
+                    id,
+                    self.opts.clone(),
+                );
+
+                let handle = session.handle();
+
+                let cancellation = self.cancellation.child_token();
+                let _ = TaskCenter::spawn(
+                    TaskKind::Background,
+                    "ingestion-partition-session",
+                    async move {
+                        session.start(cancellation).await;
+                        Ok(())
+                    },
+                );
+
+                handle
+            })
+            .value()
+            .clone()
+    }
+}
+
+/// Manager that owns all partition sessions and caches their handles.
+#[derive(Clone)]
+pub struct SessionManager<T> {
+    inner: Arc<SessionManagerInner<T>>,
+}
+
+impl<T> SessionManager<T> {
+    /// Creates a new session manager with optional overrides for session behaviour.
+    pub fn new(
+        networking: Networking<T>,
+        partition_routing: PartitionRouting,
+        opts: Option<SessionOptions>,
+    ) -> Self {
+        let inner = SessionManagerInner {
+            networking,
+            partition_routing,
+            opts: opts.unwrap_or_default(),
+            handles: Default::default(),
+            cancellation: CancellationToken::new(),
+        };
+
+        Self {
+            inner: Arc::new(inner),
+        }
+    }
+
+    pub fn partition_routing(&self) -> &PartitionRouting {
+        &self.inner.partition_routing
+    }
+
+    pub fn networking(&self) -> &Networking<T> {
+        &self.inner.networking
+    }
+}
+
+impl<T> SessionManager<T>
```

Comment:
Should we add a drop implementation that cancels `self.inner.cancellation`? This will make sure that all spawned tasks belonging to this session, will stop.